test/core/xds/xds_client_ads_stream_wait_test.cc

```diff
@@ -0,0 +1,93 @@
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <memory>
+#include <string_view>
+#include <utility>
+
+#include "absl/status/status.h"
+#include "absl/types/optional.h"
+#include "gtest/gtest.h"
+
+#include <grpc/grpc.h>
+
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/proto/grpc/testing/xds/v3/discovery.pb.h"
+#include "test/core/util/test_config.h"
+#include "test/core/xds/xds_client_test_lib.h"
+
+namespace grpc_core {
+namespace testing {
+namespace {
+
+using XdsClientNotifyWatchersDone = XdsClientTestBase;
+
+TEST_F(XdsClientNotifyWatchersDone, Basic) {
+  InitXdsClient();
+  // Start a watch for "foo1".
+  auto watcher = StartFooWatch("foo1");
+  // Watcher should initially not see any resource reported.
+  EXPECT_FALSE(watcher->HasEvent());
+  // XdsClient should have created an ADS stream.
+  auto stream = WaitForAdsStream();
+  ASSERT_TRUE(stream != nullptr);
+  // XdsClient should have sent a subscription request on the ADS stream.
+  auto request = WaitForRequest(stream.get());
+  ASSERT_TRUE(request.has_value());
+  CheckRequest(*request, XdsFooResourceType::Get()->type_url(),
+               /*version_info=*/"", /*response_nonce=*/"",
+               /*error_detail=*/absl::OkStatus(),
+               /*resource_names=*/{"foo1"});
+  CheckRequestNode(*request);  // Should be present on the first request.
+  // Send a response.
+  stream->SendMessageToClient(
+      ResponseBuilder(XdsFooResourceType::Get()->type_url())
+          .set_version_info("1")
+          .set_nonce("A")
+          .AddFooResource(XdsFooResource("foo1", 6))
+          .Serialize());
+  // XdsClient should have delivered the response to the watcher.
+  auto resource = watcher->WaitForNextResourceAndHandle();
```

Comment:
Here we need a way to control when the handle is returned, so that we can check that the `XdsClient` did not start a new read.  Then we can free the handle and verify that it *did* start a new read.

Reply:
I check it before releasing the handle:
```
  EXPECT_EQ(stream->read_count(), 1);
  resource = absl::nullopt;
  EXPECT_EQ(stream->read_count(), 2);
```